code/datums/components/full_auto_fire.dm

```diff
@@ -0,0 +1,357 @@
+	
+
+/datum/component/automatic_fire/proc/on_mouse_drag(client/source, atom/src_object, atom/over_object, turf/src_location, turf/over_location, src_control, over_control, params)
+	if(isnull(over_location)) //This happens when the mouse is over an inventory or screen object, or on entering deep darkness, for example.
+		var/list/modifiers = params2list(params)
+		var/new_target = params2turf(modifiers["screen-loc"], get_turf(source.eye), source)
+		if(!new_target)
+			target = get_turf(target) //If previous target wasn't a turf, let's turn it into one to avoid locking onto a potentially moving target.
+			CRASH("on_mouse_drag failed to get the turf under screen object [over_object.type]")
+		target = new_target
+		return
+	target = over_object
+	mouse_parameters = params
+
+
+/datum/component/automatic_fire/proc/process_shot()
+	if(autofire_stat != AUTOFIRE_STAT_FIRING)
+		return
+	switch(get_dist(shooter, target))
+		if(-1 to 0)
+			target = get_step(shooter, shooter.dir) //Shoot in the direction faced if the mouse is on the same tile as we are.
+		if(8 to INFINITY) //Can technically only go as far as 127 right now.
+			stop_autofiring() //Elvis has left the building.
+			return FALSE
+	shooter.face_atom(target)
+	if(SEND_SIGNAL(parent, COMSIG_GUN_SHOT_AUTOFIRE, target, shooter, mouse_parameters, ++shots_fired) & COMSIG_GUN_SHOT_AUTOFIRE_SUCCESS)
+		return TRUE
+	stop_autofiring()
+	return FALSE
+
+
+/datum/component/automatic_fire/proc/process_burst()
+	set waitfor = FALSE
+	if(autofire_stat != AUTOFIRE_STAT_FIRING)
+		return
+	var/current_burst = ++bursts_fired
+	for(var/i in 1 to shots_to_fire)
+		if(current_burst != bursts_fired)
+			return
+		if(!process_shot())
+			return
+		stoplag(burstfire_shot_delay)
+	
+
+/datum/component/automatic_fire/proc/itemgun_equipped(datum/source, mob/shooter, slot)
+	switch(slot)
+		if(SLOT_L_HAND, SLOT_R_HAND)
+			autofire_on(shooter.client)
+		else
+			switch(autofire_stat)
+				if(AUTOFIRE_STAT_ALERT, AUTOFIRE_STAT_FIRING)
+					autofire_off()
+
+
+/datum/component/automatic_fire/proc/modify_firedelay(datum/source, new_delay)
+	autofire_shot_delay = new_delay
+
+/datum/component/automatic_fire/proc/modify_burst_delay(datum/source, new_delay)
+	burstfire_shot_delay = new_delay
+
+/datum/component/automatic_fire/proc/modify_burst_amount(datum/source, new_amount)
+	shots_to_fire = new_amount
+
+// Gun procs.
+
+/obj/item/weapon/gun/proc/on_autofire_start(mob/living/shooter)
+	if(!able_to_fire(shooter) || gun_on_cooldown(shooter))
+		return FALSE
+	return TRUE
+
+
+/obj/item/weapon/gun/proc/on_autofire_stop(shots_fired)
+	if(shots_fired > 1) //If it was a burst, apply the burst cooldown.
+		extra_delay = min(extra_delay+(burst_delay*2), fire_delay*3)
+
+
+/obj/item/weapon/gun/proc/autofire_bypass_check(datum/source, client/clicker, atom/target, turf/location, control, params)
+	if(clicker.mob.get_active_held_item() != src)
+		return COMSIG_AUTOFIRE_ONMOUSEDOWN_BYPASS
+	return NONE //No flags means success.

```

Comment:
You don't need to return NONE on your own
```suggestion
```

Reply:
This was meant more for clarity, but sure, it's functionally the same.